### PR TITLE
feat: Add immiscible multiphase flow package

### PR DESCRIPTION
**Pull request by CusiniM**
Implementation of a multiphase immiscible solver for compressible fluids with viscous, gravity and capillary forces:

- Included new `ImmiscibleMultiphaseFlow` solver and associated kernels (currently limited to two phases)
- Standard and total mass formulations
- Cell-centered Dirichlet and flux boundary conditions
- Included new `TwoPhaseFluid` model for tabular data input
- Direct and iterative linear solver compatibility
- GPU compatibility
- Set of input files with validation cases
- Unit tests and documentation provided

### CODE REVIEW

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlowFields.hpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file CompositionalMultiphaseBaseFields.hpp
+ */
+
+#ifndef GEOS_PHYSICSSOLVERS_FLUIDFLOW_COMPOSITIONALMULTIPHASEBASEFIELDS_HPP_
+#define GEOS_PHYSICSSOLVERS_FLUIDFLOW_COMPOSITIONALMULTIPHASEBASEFIELDS_HPP_
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlowFields.hpp]:
```suggestion
#ifndef GEOS_PHYSICSSOLVERS_FLUIDFLOW_IMMISCIBLEMULTIPHASEFIELDS_HPP_
#define GEOS_PHYSICSSOLVERS_FLUIDFLOW_IMMISCIBLEMULTIPHASEFIELDS_HPP_
```

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlowFields.hpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+               LEVEL_0,
+               NO_WRITE,
+               "Phase CFL number" );
+}
+
+}
+
+}
+
+#endif // GEOS_PHYSICSSOLVERS_FLUIDFLOW_COMPOSITIONALMULTIPHASEBASEFIELDS_HPP_
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlowFields.hpp]:
```suggestion
#endif // GEOS_PHYSICSSOLVERS_FLUIDFLOW_IMMISCIBLEMULTIPHASEFIELDS_HPP_
```

File: .vscode-codespaces/launch.json [hunk 3]
```diff
@@ -9,8 +9,8 @@
             "name": "Debug geos",
             "type": "cppdbg",
             "request": "launch",
-            "program": "${userHome}/geos-build/bin/geosx",
-            "args": ["-i", "${workspaceFolder}/inputFiles/simplePDE/10x10x10Hex_LaplaceFEM_smoke.xml", "-o", "${userHome}/geos-output"],
+            "program": "${userHome}/two-phase/GEOS/install-ubuntu22-debug/bin/geosx",
+            "args": ["-i", "/home/rpiazza/two-phase/GEOS/inputFiles/immiscibleMultiphaseFlow/immiscible_2phaseFlow_1d.xml", "-o", "${userHome}/geos-output"],
```
Comment (CusiniM) [inline .vscode-codespaces/launch.json]: Remove these changes.

File: .vscode-codespaces/settings.json
Comment (CusiniM) [inline .vscode-codespaces/settings.json:1]: Remove all changes from this file.

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.hpp [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+  ///@{
+
+  /// string name to use for this class in the catalog
+  static constexpr auto m_catalogNameString = "TwoPhaseFluid";
+
+  /**
+   * @brief Static catalog string
+   * @return A string that is used to register/lookup this class in the registry
+   */
+  static std::string catalogName() { return m_catalogNameString; }
```
Comment (dkachuma) [inline src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.hpp]:
```suggestion
  static string catalogName() { return "TwoPhaseFluid"; }
```
I would just hard-code this here instead of having the static member variable.

File: src/coreComponents/unitTests/constitutiveTests/testTwoPhaseFluid.cpp [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+#include "unitTests/constitutiveTests/constitutiveTestHelpers.hpp"
+
+
+using namespace geos;
+using namespace geos::testing;
+using namespace geos::constitutive;
+using namespace geos::dataRepository;   /// Only for group definition
+
+
+static constexpr char const * tableContentPhase0 = "# Pg(Pa) Dens(kg/m3) Visc(Pa.s)\n"
```
Comment (dkachuma) [inline src/coreComponents/unitTests/constitutiveTests/testTwoPhaseFluid.cpp:32]: Normally density should increase with pressure. Otherwise a negative compressibility might cause issues (in actual simulations).

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  string const densityTableName = getName() + "DensityPhase" + GEOS_FMT( "{}", ip );
+  string const viscosityTableName = getName() +  "ViscosityPhase" + GEOS_FMT( "{}", ip );
+  m_densityTableNames.emplace_back( densityTableName );
+  m_viscosityTableNames.emplace_back( viscosityTableName );
+
+  FunctionManager & functionManager = FunctionManager::getInstance();
+
+  TableFunction & tableDensity =
+    dynamicCast< TableFunction & >( *functionManager.createChild( "TableFunction", densityTableName ) );
+  tableDensity.setTableCoordinates( pressureCoords, { units::Pressure } );
```
Comment (dkachuma) [inline src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp:146]: Do we need to check that `pressureCoords` is increasing? Or perhaps the `TableFunction` object already checks this?

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+}
+
+
+void TwoPhaseFluid::postInputInitialization()
+{
+  ConstitutiveBase::postInputInitialization();
+
+  // Input relationships can be provided either as text files or TableFunctions.
+  m_tableFiles.empty() ? readInputDataFromTableFunctions() : readInputDataFromFileTableFunctions();
+
```
Comment (dkachuma) [inline src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp:106]: Once the tables are created/read we need to check them for monotonicity. The density has to increase with pressure. Otherwise we will have a fluid with negative compressibility which might affect convergence down the line.

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 8]
```diff
@@ -0,0 +1,997 @@
...
+    CoupledSolidBase const & porousMaterial = getConstitutiveModel< CoupledSolidBase >( subRegion, solidName );
+    porousMaterial.initializeState();
+
+    // 4.3 Initialize/update the relative permeability model using the initial phase volume fraction
+    //     This is needed to handle relative permeability hysteresis
+    //     Also, initialize the fluid model
+    //
+    // Note:
+    // - This must be called after updatePhaseVolumeFraction
+    // - This step depends on phaseVolFraction
```
Comment (dkachuma) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:426]: I didn't see the phase volume fraction update. Or maybe, are saturations primary variables in this solver? I can't quite remember.

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp (hunk 6 above)
Comment (bd713) [reply src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp:146]: Yes, the `TableFunction` has already a built-in check to ensure coordinates are strictly increasing

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp (hunk 7 above)
Comment (bd713) [reply src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp:106]: Thanks for the heads up. I will add a `checkTableConsistency` as I have seen it implemented in other fluid models.

File: src/coreComponents/unitTests/constitutiveTests/testTwoPhaseFluid.cpp (hunk 5 above)
Comment (bd713) [reply src/coreComponents/unitTests/constitutiveTests/testTwoPhaseFluid.cpp:32]: Duly noted regarding the monotonicity. I will update the test with more "realistic" setups.

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+                 InputError );
+
+  // Read line-by-line until eof
+  string str;
+  while( std::getline( is, str ) )
+  {
+    // Remove whitespace and end-of-line characters, if any
+    str = stringutilities::trim( str, " \r" );
+
+    // Remove # and -- (Eclipse-style) comments
```
Comment (paveltomin) [inline src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp]: remove "Eclipse"

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+     resizeFields( 0, 0 );
+
+     // set labels on array wrappers for plottable fields
+     setLabels();
+   */
+}
+
+
+void
+TwoPhaseFluid::readTable( string const & fileName,
```
Comment (paveltomin) [inline src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp]: this function should be already implemented somewhere?

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 11]
```diff
@@ -0,0 +1,1277 @@
...
+
+  immiscibleMultiphaseKernels::
+    PhaseMobilityKernelFactory::
+    createAndLaunch< parallelDevicePolicy<> >( m_numPhases,
+                                               dataGroup,
+                                               fluid,
+                                               relperm );
+}
+
+void ImmiscibleMultiphaseFlow::initializeFluidState( MeshLevel & mesh,
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:371]: check https://github.com/GEOS-DEV/GEOS/pull/3393, i hope it is not going to be hard to merge

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 12]
```diff
@@ -0,0 +1,1277 @@
...
+    FieldIdentifiers fieldsToBeSync;
+    fieldsToBeSync.addElementFields( fields, regionNames );
+
+    CommunicationTools::getInstance().synchronizeFields( fieldsToBeSync, mesh, domain.getNeighbors(), true );
+  } );
+
+}
+
+
+void ImmiscibleMultiphaseFlow::updatePhaseVolumeFraction( ElementSubRegionBase & subRegion ) const
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp]: i would not call it the same name as in compositional solver, i sounds a bit confusing

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 13]
```diff
@@ -0,0 +1,1277 @@
...
+      // update all fluid properties
+      updatePhaseVolumeFraction( subRegion );
+      updateFluidState( subRegion );
+    } );
+  } );
+}
+
+real64 ImmiscibleMultiphaseFlow::setNextDt( const geos::real64 & currentDt, geos::DomainPartition & domain )
+{
+  return PhysicsSolverBase::setNextDt( currentDt, domain );
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:1340]: TODO pressure and saturation change time step selector

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp [hunk 14]
```diff
@@ -0,0 +1,1420 @@
...
+        {
+          // calculate quantities on primary connected cells
+          for( integer ke = 0; ke < 2; ++ke )
+          {
+            // density
+            real64 const density  = m_dens[seri[ke]][sesri[ke]][sei[ke]][0][ip];         // r = rho1 || rho2
+            real64 const dDens_dP = m_dDens_dPres[seri[ke]][sesri[ke]][sei[ke]][0][ip][Deriv::dP];  // dr/dP = dr1/dP1 || dr2/dP
+
+            // average density and derivatives
+            densMean[ip] += 0.5 * density;          // rho = (rho1 + rho2) / 2
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: this treatment was recently modified, see https://github.com/GEOS-DEV/GEOS/pull/3337

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp [hunk 15]
```diff
@@ -0,0 +1,1420 @@
...
+          }
+
+          //***** calculation of flux *****
+
+          // compute potential difference
+          real64 potScale = 0.0;
+          real64 dPresGrad_dTrans = 0.0;
+          real64 dGravHead_dTrans = 0.0;
+          real64 dCapGrad_dTrans = 0.0;
+          int signPotDiff[2] = {1, -1};
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: i don't think you need this sign thing

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp [hunk 16]
```diff
@@ -0,0 +1,1420 @@
...
+          // *** upwinding ***
+
+          // compute potential gradient
+          real64 potGrad = presGrad[ip] - gravHead[ip]; // DPhi = T (P1 - P2) - T rho g (z1 - z2)
+          if( m_hasCapPressure )
+          {
+            potGrad += capGrad[ip]; // DPhi = T (P1 - P2) - T rho g (z1 - z2) + T (-Pc1 + Pc2)
+          }
+
+          // compute upwinding tolerance
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp:532]: compositional is not doing that trick but ok to keep it here as long as it works fine

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp [hunk 17]
```diff
@@ -0,0 +1,1420 @@
...
+                       flowAccessors, fluidAccessors, capPressureAccessors, permAccessors,
+                       dt, localMatrix, localRhs, hasCapPressure, useTotalMassEquation );
+    kernelType::template launch< POLICY >( stencilWrapper.size(), kernel );
+  }
+};
+
+
+/******************************** AccumulationKernel ********************************/
+
+static constexpr real64 minDensForDivision = 1e-10;
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: rename/remove?

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp [hunk 18]
```diff
@@ -0,0 +1,1420 @@
...
+   * @param[inout] stack the stack variables
+   * @param[in] phaseAmountKernelOp the function used to customize the kernel
+   */
+  template< typename FUNC = NoOpFunc >
+  GEOS_HOST_DEVICE
+  void computeAccumulation( localIndex const ei,
+                            StackVariables & stack,
+                            FUNC && phaseAmountKernelOp = NoOpFunc{} ) const
+  {
+      int signPotDiff[2] = {1, -1};
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: confusing name

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp [hunk 19]
```diff
@@ -0,0 +1,1420 @@
...
+  template< typename FUNC = NoOpFunc >
+  GEOS_HOST_DEVICE
+  void compute( localIndex const ei,
+                FUNC && phaseMobilityKernelOp = NoOpFunc{} ) const
+  {
+    using Deriv = immiscibleFlow::DerivativeOffset;
+
+    arraySlice1d< real64, immiscibleFlow::USD_PHASE - 1 > const phaseMob = m_phaseMob[ei];
+    arraySlice2d< real64, immiscibleFlow::USD_PHASE_DS - 1 > const dPhaseMob = m_dPhaseMob[ei];
+    int sign[2] = {1, -1};
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: can you avoid using sign here and use fixed second index for `m_dPhaseRelPerm_dPhaseVolFrac` instead, like:
`real64 const dRelPerm_dS = m_dPhaseRelPerm_dPhaseVolFrac[ei][0][ip][0];`

**Comment by paveltomin** [comment]
Overall looks pretty good, just some minor things here and there, cleanup and remove commented code

File: src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/ImmiscibleMultiphaseFVM.hpp [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+{
+public:
+  /**
+   * @brief Constructor.
+   * @param numComponentsPerField array with number of components for each field
+   */
+  explicit ImmiscibleMultiphaseFVM( arrayView1d< int const > const & numComponentsPerField )
+    : MGRStrategyBase( LvArray::integerConversion< HYPRE_Int >( numComponentsPerField[0] ) )
+  {
+    // Level 0: eliminate last density which corresponds to the volume constraint equation
```
Comment (victorapm) [inline src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/ImmiscibleMultiphaseFVM.hpp]: These comments can be removed

**Review by victorapm — APPROVED** [review]
I just looked at MGR changes, looks good!

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp (hunk 9 above)
Comment (rpiazza87) [reply src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp]: Done. See next comment.

File: src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp (hunk 10 above)
Comment (rpiazza87) [reply src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp]: The function has been removed and we are now using a previously implemented version.

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp (hunk 12 above)
Comment (rpiazza87) [reply src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp]: Renamed to updateVolumeConstraint

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp (hunk 13 above)
Comment (rpiazza87) [reply src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:1340]: Implemented new function:
`real64 ImmiscibleMultiphaseFlow::setNextDtBasedOnStateChange(real64 const & currentDt, DomainPartition & domain) `

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp (hunk 14 above)
Comment (rpiazza87) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: New treatment implemented in kernel:
 `void computeFlux( localIndex const iconn, StackVariables & stack, FUNC && kernelOp = NoOpFunc{} ) const`

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp (hunk 17 above)
Comment (rpiazza87) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: Removed line of code

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp (hunk 18 above)
Comment (rpiazza87) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: Renamed to just `sign[2]`

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp (hunk 19 above)
Comment (rpiazza87) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: The mobility derivative matrix we are currently using is diagonal. This implementation handles missing off-diagonal entry.

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp (hunk 15 above)
Comment (rpiazza87) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: Keeping it for now. We can think of a more elegant implementation in the future.

Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: i think it is just wrong, why no just this:
```
            presGrad[ip] += trans[ke] * pressure;
...
            dPresGrad_dP[ip][ke] += dTrans_dP[ke] * pressure;

```
?

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp (hunk 14 above)
Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: sorry we had to revert that new treatment and make it an option, see https://github.com/GEOS-DEV/GEOS/pull/3467
somehow old treatment works much better for spe11 and it is not a good time to spoil it

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp (hunk 15 above)
Comment (rpiazza87) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: If we understand correctly, it's because we need a pressure difference instead of a single pressure value:

Here presGrad = T * (P1-P2)
When we take the derivative with respect to the two pressures we will need: dPresGrad_dP = {dT/dP1 * (P1-P2) + T, dT/dP1 * (P1-P2) - T}. At the last line of code shown dPresGrad_dP = {T, -T} already.
So the second term on the right-hand side needs to be a (P1-P2), instead of just a single pressure. We used the signPotDiff variable to assist taking the difference of the 2 pressures.

Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: i am still confused
the code is more like
`presGrad = T1*P1 + T2*P2`
right?
the rest should follow the same notation/logic
maybe @CusiniM or @ryar9534 can have a look and check if i am wrong

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp (hunk 14 above)
Comment (rpiazza87) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: We adapted the code to make it optional in the immiscible solver as well, using the parameter `m_checkPhasePresenceInGravity`

File: src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp (hunk 15 above)
Comment (rpiazza87) [reply src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp]: Someone else can check our understanding, but it's because in
 `presGrad[ip] += trans[ke] * pressure;    // DPv = T (P1 - P2)` 
we are using the transmissibilities that come from the `computeWeights (CellElementStencilTPFA.hpp)` function like this:
`weight[0][ke] = m_transMultiplier[iconn] * value * (ke == 0 ? 1 : -1);`
So T2 = -T1

**Comment by matteofrigo5** [comment]
I'd remove the file `host-configs/Stanford/ubuntu22.cmake` before merging this branch into the develop branch.

**Review by paveltomin — APPROVED** [review]

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 21]
```diff
@@ -0,0 +1,1345 @@
...
+void ImmiscibleMultiphaseFlow::postInputInitialization()
+{
+  FlowSolverBase::postInputInitialization();
+}
+
+void ImmiscibleMultiphaseFlow::registerDataOnMesh( Group & meshBodies )
+{
+  FlowSolverBase::registerDataOnMesh( meshBodies );
+
+  // DomainPartition const & domain = this->getGroupByPath< DomainPartition >( "/Problem/domain" );
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:103]: remove?

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 22]
```diff
@@ -0,0 +1,1345 @@
...
+
+      subRegion.registerField< phaseMobility >( getName() ).
+        reference().resizeDimension< 1 >( m_numPhases );
+
+      subRegion.registerField< dPhaseMobility >( getName() ).
+        reference().resizeDimension< 1, 2 >( m_numPhases, m_numPhases ); // dP, dS
+
+    } );
+
+    // FaceManager & faceManager = mesh.getFaceManager();
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:177]: remove?

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 23]
```diff
@@ -0,0 +1,1345 @@
...
+}
+
+void ImmiscibleMultiphaseFlow::initializePreSubGroups()
+{
+  m_linearSolverParameters.get().mgr.strategy = LinearSolverParameters::MGR::StrategyType::immiscibleMultiphaseFVM;
+
+  FlowSolverBase::initializePreSubGroups();
+
+  DomainPartition & domain = this->getGroupByPath< DomainPartition >( "/Problem/domain" );
+  // ConstitutiveManager const & cm = domain.getConstitutiveManager();
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:216]: remove

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 24]
```diff
@@ -0,0 +1,1345 @@
...
+
+  // // 1. Validate various models against each other (must have same phases and components)
+  // validateConstitutiveModels( domain );
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:219]: check and remove if not needed

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 25]
```diff
@@ -0,0 +1,1345 @@
...
+
+  arrayView1d< real64 const > const volume = subRegion.getElementVolume();
+  arrayView2d< real64 const > const porosity = solid.getPorosity();
+
+  arrayView2d< real64 const, immiscibleFlow::USD_PHASE > const phaseVolFrac= subRegion.getField< fields::immiscibleMultiphaseFlow::phaseVolumeFraction >();
+  arrayView3d< real64 const, multifluid::USD_PHASE > phaseDens = fluid.phaseDensity();
+  arrayView2d< real64, immiscibleFlow::USD_PHASE > const phaseMass = subRegion.getField< fields::immiscibleMultiphaseFlow::phaseMass >();
+
+  // Might be needed for geomechanics????? if so, need to change the accumulation as well?
+  //arrayView1d< real64 > const deltaVolume = subRegion.getField< fields::flow::deltaVolume >();
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:339]: this is used for geomechanics with fractures yes, see https://github.com/GEOS-DEV/GEOS/pull/3228/files#diff-44d45911b0c4b17989e2a0ba29b844cb0ee4975f16329ff612825db0b269d55b
not really needed in your PR at this point

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 26]
```diff
@@ -0,0 +1,1345 @@
...
+{
+  GEOS_MARK_FUNCTION;
+
+  // apply pressure boundary conditions.
+  applyDirichletBC( time_n, dt, dofManager, domain, localMatrix.toViewConstSizes(), localRhs.toView() );
+
+// apply flux boundary conditions
+  applySourceFluxBC( time_n, dt, dofManager, domain, localMatrix.toViewConstSizes(), localRhs.toView() );
+
+  //   for( localIndex row = 0; row < localMatrix.toViewConstSizes().numRows(); ++row )
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:689]: remove, there is a general way to print out matrix and rhs in base solver

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 27]
```diff
@@ -0,0 +1,1345 @@
...
+void ImmiscibleMultiphaseFlow::applyDirichletBC( real64 const time_n,
+                                                 real64 const dt,
+                                                 DofManager const & dofManager,
+                                                 DomainPartition & domain,
+                                                 CRSMatrixView< real64, globalIndex const > const & localMatrix,
+                                                 arrayView1d< real64 > const & localRhs ) const
+{
+  GEOS_MARK_FUNCTION;
+
+  // // Only validate BC at the beginning of Newton loop
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:718]: restore or remove

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 28]
```diff
@@ -0,0 +1,1345 @@
...
+          }
+        }
+        else
+        {
+          globalIndex const compMassBalanceRow = dofNumber[ei] - rankOffset + fluidPhaseId;
+          localRhs[compMassBalanceRow] += rhsValue;
+        }
+      } );
+
+      // SourceFluxStatsAggregator::forAllFluxStatWrappers( subRegion, fs.getName(),
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:963]: restore?

File: src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp [hunk 29]
```diff
@@ -0,0 +1,1345 @@
...
+void ImmiscibleMultiphaseFlow::applySystemSolution( DofManager const & dofManager,
+                                                    arrayView1d< real64 const > const & localSolution,
+                                                    real64 const scalingFactor,
+                                                    real64 const dt,
+                                                    DomainPartition & domain )
+{
+  GEOS_UNUSED_VAR( dt );
+
+  DofManager::CompMask pressureMask( m_numDofPerCell, 0, 1 );
+  //DofManager::CompMask PhaseMask( m_numDofPerCell, 1, m_numPhases );
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp:1068]: remove if not needed

File: src/coreComponents/linearAlgebra/interfaces/hypre/HypreMGR.hpp [hunk 30]
```diff
@@ -213,6 +213,7 @@ class MGRStrategyBase
 #if GEOS_USE_HYPRE_DEVICE == GEOS_USE_HYPRE_CUDA || GEOS_USE_HYPRE_DEVICE == GEOS_USE_HYPRE_HIP
     GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetAggInterpType( solver.ptr, hypre::getAMGAggressiveInterpolationType( LinearSolverParameters::AMG::AggInterpType::modifiedExtendedE ) ) );
     GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetCoarsenType( solver.ptr, hypre::getAMGCoarseningType( LinearSolverParameters::AMG::CoarseningType::PMIS ) ) );
+    GEOS_LAI_CHECK_ERROR( HYPRE_BoomerAMGSetAggNumLevels( solver.ptr, 0 ) );
```
Comment (CusiniM) [inline src/coreComponents/linearAlgebra/interfaces/hypre/HypreMGR.hpp:216]: can you remove this and create a PR only with this change? @victorapm we should test this. Unless you know why this would impact GPU performance so much.
